src/charm.py

```diff
@@ -252,6 +326,198 @@ def _on_nginx_route_broken(self, _: Any) -> None:
+        if not tls_certificates_relation:
+            event.fail("Certificates relation not created.")
+            return
+        if tls_certificates_relation.data[self.unit]["certificate"]:
+            event.set_results(
+                {
+                    "certificate": tls_certificates_relation.data[self.unit]["certificate"],
+                    "ca": tls_certificates_relation.data[self.unit]["ca"],
+                    "chain": tls_certificates_relation.data[self.unit]["chain"],
+                }
+            )
+        else:
+            event.fail("Certificate not available")
+
+    def _on_certificates_relation_created(self, event: RelationCreatedEvent) -> None:
+        """Handle the TLS Certificate relation created event.
+
+        Args:
+            event: The event that fires this method.
+        """
+        tls_certificates_relation = self.model.get_relation("certificates")
+        if not tls_certificates_relation:
+            self.unit.status = WaitingStatus("Waiting for peer relation to be created")
+            event.defer()
+            return
+        private_key_password = self._tls.generate_password().encode()
+        private_key = generate_private_key(password=private_key_password)
+        private_key_dict = {"password": private_key_password.decode(), "key": private_key.decode()}
+        try:
+            secret = self.model.get_secret(label="private-key")
+            secret.set_content(private_key_dict)
+        except ModelError:
+            secret = self.app.add_secret(content=private_key_dict, label="private-key")
+        tls_certificates_relation.data[self.unit].update(
+            {
+                "private_key_password": private_key_password.decode(),
+                "private_key": private_key.decode(),
+            }
+        )
+
+    def _on_certificates_relation_joined(self, event: RelationJoinedEvent) -> None:
+        """Handle the TLS Certificate relation joined event.
+
+        Args:
+            event: The event that fires this method.
+        """
+        tls_certificates_relation = self.model.get_relation("certificates")
+        if not tls_certificates_relation:
+            self.unit.status = WaitingStatus("Waiting for peer relation to be created")
+            event.defer()
+            return
+
+        secret = self.model.get_secret(label="private-key")
+        secret.grant(tls_certificates_relation)
+        relation = self._get_relation()
+        if relation is None:
+            self._cleanup()
+            self.unit.status = WaitingStatus("waiting for relation")
+            return
+        definition = self._get_definition_from_relation(relation)
+        subject = definition.service_hostname if definition is not None else self.model.name
+        csr = generate_csr(
+            private_key=secret.get_content()["key"].encode(),
+            private_key_password=secret.get_content()["password"].encode(),
```

Comment:
If we are using juju secrets to store the private keys, it means that we might not be able to deploy the nginx-ingress-integrator in juju 2.9 (PS5). Maybe we should consider using the Kubernetes secrets, as the Kubernetes TLS secrets can contain the key with an empty cert and some extra fields like csr.